External/SPMmodified/xASL/xASL_im_FillNaNs.m

```diff
@@ -267,184 +267,31 @@
-                numVoxels = length(voxels6)/2;
-                robustValue = robustMedian + numVoxels*robustDiff;
-                % Now take the mean of the actual last value and the robust one
-                robustValue = mean([robustValue imThisCol(maxInd)]);
-
-                % Apply this
-                imThisCol(maxInd+1:end) = robustValue + (1:nCol-maxInd) * meanDiff;
-                % Put this column back in the image
-                imCol(:, ColumnIs) = imThisCol;
-            end
-    
-            %% 4. Reshape 2D back to 3D
-            switch(DimensionIs)
-                case 1
-                    IM = reshape(imCol, [nX,nY,nZ]);
-                case 2
-				    % For 2D matrix and 3D matrix, the reshaping back is done differently
-				    if ndims(IM)<3
-					    IM = shiftdim(reshape(imCol,[nY,nX]),1);
-				    else
-					    IM = shiftdim(reshape(imCol,[nY,nZ,nX]),2);
-				    end
-                case 3
-                    IM = reshape(imCol', [nX,nY,nZ]);
-            end % switch
-        end % for DimensionIs
-    end % sum(isnan(IM))
-end
-
-% Any remaining NaNs had to be inside, so we interpolate them (setting them
-% to zeros creates artifacts!)
-IM = xASL_im_ExtrapolateSmoothOverNaNs(IM, 0);
-
-end
+%% Extrapolation of values based on the edge values 
+% Calculate the distance map from the non-Nan towards NaN
+[~, distY, distX, distZ] = xASL_im_DistanceTransform(~maskNaN);
 
+% We calculate the meshgrid
+[gridX, gridY, gridZ] = ndgrid(1:size(IM, 1), 1:size(IM, 2), 1:size(IM, 3));
 
-function [diffIs, minInd, maxInd, thresholdIs] = xASL_im_Extrapolate_FixEdges(imThisCol, bSide)
-%xASL_im_Extrapolate_FixEdges Check if the edges of the non-NaN voxels within a column are outliers
-% from the distribution of differences between voxels. If they are outliers, they are probably interpolation 
-% errors from the SPM software, and we want to ignore them.
-%
-% input
-% imThisCol - vector of current column (REQUIRED)
-% bSide     - boolean specifying if we check the start/first non-NaN voxel (0)
-% %           or the end/last non-NaN voxel (1). (REQUIRED)
-
-
-minInd = find(~isnan(imThisCol), 1, 'first'); % Get the index of the first non-value
-maxInd = find(~isnan(imThisCol), 1, 'last'); % Get the index of the last non-value
-midInd = floor(minInd+(maxInd-minInd)/2); % Get the index of the middle voxel
+% We calculate the coordinates of the border voxel
+borderX = distX + gridX;
+borderY = distY + gridY;
+borderZ = distZ + gridZ;
 
-% Mean difference, from the middle index to the index of the last value
-% Calculated by their difference with their neighboring voxel            
-
-if ~bSide % start/first
-    diffIs = imThisCol(minInd:midInd) - imThisCol(minInd+1:midInd+1);
-else
-    diffIs = imThisCol(midInd:maxInd) - imThisCol(midInd-1:maxInd-1);
+for iDim = 1:3
+	% To each NaN voxel, we assign the border value in the original image
```

Comment:
You forgot to use the NaNmask here right? Now you reset all voxels

Reply:
Haha - no I didn't ;) That's the beauty of the solution. For NaNs, we will reassign the new value. For non-NaNs, distX,Y,Z is zero, so we reassign the map with the exactly same value.

So this is exactly what we functionally want and it's much faster and more elegant.
I added a comment in the code.

Reply:
OK but then I don't understand. gridX gives the original coordinate of the matrix, not the original values of the non-nans right? And we need those to stay the original values of the non-nans (but perhaps that doesn't matter because you only change the values inside the MaskNan below)

Reply:
GridX gives the coordinate for each voxel - each voxel has it's own coordinate in gridX.
GridX + distanceX gives the coordinate of the closest border voxel for each voxel. So non-NaNs get their own coordinates.

I update the entire IM.  But updating IM with values in gridX+distanceX will give border value to NaNs and will do nothing to non-NaNs. This is perfectly safe, secure, does what it should exactly and it's computationally much faster. So no reason to update only MaskNan, it's better to update all.

Reply:
How I read it (please correct me where I'm wrong so I understand, I trust that you're doing it right, just want to understand :)

1. GridX gives the coordinate for each voxel (which is wrong for non-NaNs because they should not have their own coordinates but their original non-linearly created values).
-> this is probably something I don't understand correctly?

2. We don't want to give border-values to non-NaNs, these need to be the smoothed border-values (more robust, less non-linear)
-> this is something I think you should add

Reply:
I've updated my last comment. I thought that this is clear because of the way gridX is created...

GridX is using the coordinates within the transformation matrix. So 1,2,3,4
Values of IM are the coordinates of the transformation in real world space. So 0.5mm, 1mm, 1.5mm etc.

Reply:
But that it what I didn't understand. IM should not be 0.5mm 1mm 1.5mm, it should only be that for the voxels in NaNmask. The voxels outside NaNmask should stay the same as they were (the non-linear flowfield), not being replaced by "identity". Perhaps I just didn't understand your code enough. I guess you can resolve and try to explain me in the Whatsapp conversation :)